src/main/java/org/zalando/nakadi/validation/SchemaCompatibilityChecker.java

```diff
@@ -0,0 +1,116 @@
+package org.zalando.nakadi.validation;
+
+import com.google.common.collect.Lists;
+import org.everit.json.schema.ArraySchema;
+import org.everit.json.schema.CombinedSchema;
+import org.everit.json.schema.ObjectSchema;
+import org.everit.json.schema.ReferenceSchema;
+import org.everit.json.schema.Schema;
+import org.springframework.stereotype.Component;
+import org.zalando.nakadi.validation.schema.NotSchemaConstraint;
+import org.zalando.nakadi.validation.schema.PatternPropertiesConstraint;
+import org.zalando.nakadi.validation.schema.SchemaConstraint;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Stack;
+
+@Component
+public class SchemaCompatibilityChecker {
+
+    final private List<SchemaConstraint> CONSTRAINTS = Lists.newArrayList();
+
+    public SchemaCompatibilityChecker() {
+        CONSTRAINTS.add(new NotSchemaConstraint());
+        CONSTRAINTS.add(new PatternPropertiesConstraint("patternProperties"));
+    }
+
+    public List<SchemaIncompatibility> checkConstraints(final Schema schema) {
+        final List<SchemaIncompatibility> incompatibilities = new ArrayList<SchemaIncompatibility>();
+
+        recursiveCheckConstraints(schema, new Stack<String>(), incompatibilities);
+
+        return incompatibilities;
+    }
+
+    private void recursiveCheckConstraints(
+            final Schema schema,
+            final Stack<String> jsonPath,
+            final List<SchemaIncompatibility> schemaIncompatibilities) {
+
+        for (final SchemaConstraint constraint : CONSTRAINTS) {
+            final Optional<SchemaIncompatibility> incompatibility = constraint.validate(jsonPath, schema);
+            if (incompatibility.isPresent()) {
+                schemaIncompatibilities.add(incompatibility.get());
+            }
+        }
+
+        if (schema instanceof ObjectSchema) {
+            recursiveCheckConstraints((ObjectSchema) schema, jsonPath, schemaIncompatibilities);
+        } else if (schema instanceof ArraySchema) {
+            recursiveCheckConstraints((ArraySchema) schema, jsonPath, schemaIncompatibilities);
+        } else if (schema instanceof ReferenceSchema) {
+            recursiveCheckConstraints((ReferenceSchema) schema, jsonPath, schemaIncompatibilities);
+        } else if (schema instanceof CombinedSchema) {
```

Comment:
No, sorry, It will not work. It seems that in Java overloading resolution is performed at compile time. :(